src/system.rs

```diff
@@ -182,15 +193,31 @@ impl PendingTypes {
     }
 
     // helper method to update type store with all the types from this PendingTypes
-    fn update_type_store_with_all_types(&self, type_store: &mut TypeStore) -> IonSchemaResult<()> {
+    fn update_type_store_with_all_types(
+        &self,
+        type_store: &mut TypeStore,
+        isl_type_names: &HashSet<&str>,
+    ) -> IonSchemaResult<()> {
         for optional_type in &self.types_by_id {
             // return an error if any of the type in types_by_id vector is None/Unresolved
             let type_def = optional_type.to_owned().ok_or_else(|| {
                 unresolvable_schema_error_raw("Unable to load schema due to unresolvable type")
             })?;
 
             match type_def {
-                TypeDefinition::Named(named_type_def) => type_store.add_named_type(named_type_def),
+                TypeDefinition::Named(named_type_def) => {
+                    // check if the type definitions that are not yet resolved actually exists within the schema
+                    // we can use the isl_type_names to make sure if they exists, otherwise return error.
+                    if named_type_def.is_deferred_type_def()
+                        && !isl_type_names
+                            .contains(named_type_def.name().as_ref().unwrap().as_str())
+                    {
+                        return unresolvable_schema_error(
+                            "Unable to load schema due to unresolvable type",
```

Comment:
suggestion–in the error message, include the name of the type that cannot be resolved